### PR TITLE
✨ introduce runtime-openapi-gen to generate openapi spec on release 

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

This PR introduces adds the upstream `openapi-gen` to generate go openapi definitions and `runtime-openapi-gen` to generate a openapi spec for Runtime SDK hooks on release. 

Requires #6455 to get merged first.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #6330 

**TODOs**:

- [x] wait for #6455 to be merged
- [x] rebase

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -118,6 +118,16 @@ KPROMO_PKG := sigs.k8s.io/promo-tools/v3/cmd/kpromo
 CONVERSION_VERIFIER_BIN := conversion-verifier
 CONVERSION_VERIFIER := $(abspath $(TOOLS_BIN_DIR)/$(CONVERSION_VERIFIER_BIN))
 
+OPENAPI_GEN_VER := v0.0.0-20211115234752-e816edb12b65
```
Comment (sbueringer) [inline Makefile:121]: Looks like they don't have releases, but should we use the latest commit?

File: Makefile [hunk 2]
```diff
@@ -118,6 +118,16 @@ KPROMO_PKG := sigs.k8s.io/promo-tools/v3/cmd/kpromo
...
+OPENAPI_GEN_BIN := openapi-gen
+# We are intentionally using the binary without version suffix, to avoid the version
+# in generated files.
```
Comment (sbueringer) [inline Makefile:124]: I assume that's not just copy&paste and we had the same issue here as we had with conversion-gen?

File: Makefile [hunk 3]
```diff
@@ -177,10 +187,11 @@ help:  # Display this help
 ##@ generate:
 
 ALL_GENERATE_MODULES = core kubeadm-bootstrap kubeadm-control-plane
+ALL_GENERATE_OPENAPI = api/v1beta1 $(EXP_DIR)/runtime/hooks/api/v1alpha1
```
Comment (sbueringer) [inline Makefile:190]: I think I would move this to the target where it's used (probably just inline it) as we usually have the packages in the targets. (it's also only used there)

File: Makefile [hunk 4]
```diff
@@ -263,7 +275,7 @@ generate-go-deepcopy-kubeadm-control-plane: $(CONTROLLER_GEN) ## Generate deepco
 
 .PHONY: generate-go-conversions
 generate-go-conversions: ## Run all generate-go-conversions-* targets
-	$(MAKE) $(addprefix generate-go-conversions-,$(ALL_GENERATE_MODULES))
+	$(MAKE) $(addprefix generate-go-conversions-,$(ALL_GENERATE_MODULES) runtime)
```
Comment (sbueringer) [inline Makefile:278]: fyi: https://github.com/kubernetes-sigs/cluster-api/pull/6455/files#r861720835

File: Makefile [hunk 5]
```diff
@@ -803,6 +850,15 @@ $(CONVERSION_GEN): # Build conversion-gen from tools folder.
 $(CONVERSION_VERIFIER): $(TOOLS_DIR)/go.mod # Build conversion-verifier from tools folder.
 	cd $(TOOLS_DIR); go build -tags=tools -o $(BIN_DIR)/$(CONVERSION_VERIFIER_BIN) sigs.k8s.io/cluster-api/hack/tools/conversion-verifier
 
+.PHONY: $(OPENAPI_GEN)
```
Comment (sbueringer) [inline Makefile:853]: Please add the same comment here as in l.844-845

File: hack/tools/runtime-openapi-gen/main.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	if outputFileExt != ".yaml" && outputFileExt != ".json" {
+		klog.Exit("--output-file must have either 'yaml' or 'json' extension")
+	}
+
+	c := catalog.New()
+	_ = runtimehooksv1.AddToCatalog(c)
+
+	// TODO(openapi): TBD if we want to support types from "external" packages which only have GetOpenAPIDefinitions.
+	// If yes, we should add an additional flag for those.
+	// Follow-up: If we don't want to support it drop openapi generation in api/v1beta1.
```
Comment (sbueringer) [inline hack/tools/runtime-openapi-gen/main.go:58]: I think we can drop this and the next TODO comment

File: hack/tools/runtime-openapi-gen/main.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+	c.AddOpenAPIDefinitions(clusterv1.GetOpenAPIDefinitions)
```
Comment (sbueringer) [inline hack/tools/runtime-openapi-gen/main.go:59]: Looks like changes in catalog + openapi.go are missing

File: hack/tools/runtime-openapi-gen/vendored_openapi.go [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kube-openapi/pkg/common"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+// GetOpenAPIDefinitions returns OpenAPIDefinitions for external types.
+// We need them when we want to generate OpenAPI for a type embedding one of those external types.
+// Those OpenAPIDefinitions are all vendored from: https://github.com/kubernetes/kubernetes/blob/master/pkg/generated/openapi/zz_generated.openapi.go
+// Prior art: https://github.com/kubernetes-sigs/controller-tools/blob/master/pkg/crd/known_types.go.
+func GetOpenAPIDefinitions(ref common.ReferenceCallback) map[string]common.OpenAPIDefinition {
```
Comment (sbueringer) [inline hack/tools/runtime-openapi-gen/vendored_openapi.go:14]: TBD if we want to move this into the catalog.

I'm fine with both options. Opinions?

File: Makefile (hunk 1 above)
Comment (chrischdi) [reply Makefile:121]: We can, for now I did choose the same as we had on the poc branch (via go.mod), will update it 👍

File: hack/tools/runtime-openapi-gen/main.go (hunk 7 above)
Comment (chrischdi) [reply hack/tools/runtime-openapi-gen/main.go:59]: ack 🤦

File: Makefile (hunk 2 above)
Comment (chrischdi) [reply Makefile:124]: it is c&p, removed the comment and adjusted to use binary named by version.

File: Makefile (hunk 5 above)
Comment (chrischdi) [reply Makefile:853]: nope, won't do after changing to versioned binary

File: hack/tools/runtime-openapi-gen/vendored_openapi.go (hunk 8 above)
Comment (sbueringer) [reply hack/tools/runtime-openapi-gen/vendored_openapi.go:14]: On the other side the chance that anyone needs the same or a subset of our vendored OpenAPI Definitions is probably very small.

File: Makefile [hunk 9]
```diff
@@ -177,10 +187,11 @@ help:  # Display this help
 ##@ generate:
 
 ALL_GENERATE_MODULES = core kubeadm-bootstrap kubeadm-control-plane
+ALL_GENERATE_OPENAPI = api/v1beta1 exp/runtime/hooks/api/v1alpha1 exp/runtime/hooks/api/v1alpha2 exp/runtime/hooks/api/v1alpha3
```
Comment (sbueringer) [inline Makefile:190]:
```suggestion
```
Should be unused now and can be removed

**Comment by sbueringer** [comment]
No rush, but just fyi, the latest push --force didn't seem to push everything (just in case you already did implement it :) )

File: Makefile (hunk 9 above)
Comment (chrischdi) [reply Makefile]: Yes, was not able to finish it last week :-) latest commit removes it and also has now a working `make release-manifests`

**Review by ykakarap — COMMENTED** [review]
Looks like this PR is also introducing the discovery types. Dropping a note here to make sure that we wait for the Discovery PR (https://github.com/kubernetes-sigs/cluster-api/pull/6455) before we merge this.

File: exp/runtime/hooks/api/v1alpha1/groupversion_info.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	// to a Catalog.
+	catalogBuilder = &catalog.Builder{GroupVersion: GroupVersion}
+
+	// AddToCatalog adds rpc services defined in this package and their request and
+	// response types to a catalog.
+	AddToCatalog = catalogBuilder.AddToCatalog
+)
+
+func init() {
+	// Add Open API definitions for rpc method's request and response in this package
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/groupversion_info.go:40]:
```suggestion
	// Add Open API definitions for RuntimeHooks request and response types in this package
```
nit, or something similar (fyi we adjusted "rpc services" above in the underlying PR)

File: internal/runtime/catalog/openapi.go [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+			},
+			Deprecated: hookDescriptor.metadata.Deprecated,
+		},
+	}
+	path := GVHToPath(gvh, "")
+
+	// Add name parameter to operation path if necessary.
+	// TODO(openapi): TBD if we want to add the name parameter.
+	// I think it's useful as that reflects the real path which Runtime Extensions will handle.
+	// Follow-up: potentially drop Singleton field.
```
Comment (sbueringer) [inline internal/runtime/catalog/openapi.go:88]: let's just drop the todo for now and just see if we get other feedback

File: internal/runtime/catalog/openapi.go [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+	}
+
+	// Add request type to operation.
+	requestGVK, err := c.Request(gvh)
+	if err != nil {
+		return err
+	}
+	requestType, ok := c.scheme.AllKnownTypes()[requestGVK]
+	if !ok {
+		return errors.New("TODO 4")
```
Comment (sbueringer) [inline internal/runtime/catalog/openapi.go:112]: please go over the error returns and make sure we add enough context when necessary (and please replace my TODO messages :) )

File: internal/runtime/catalog/openapi.go [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+		return err
+	}
+	responseType := c.scheme.AllKnownTypes()[responseGVK]
+	if !ok {
+		return errors.New("TODO 5")
+	}
+	responseTypeName := typeName(responseType, responseGVK)
+	operation.Responses.StatusCodeResponses[http.StatusOK] = &spec3.Response{
+		// TODO(openapi): TBD do we want to handle other response codes?
+		// Not sure if we should introduce them when we don't need them.
```
Comment (sbueringer) [inline internal/runtime/catalog/openapi.go:136]: Let's just drop the TODO for now

File: Makefile (hunk 2 above)
Comment (chrischdi) [reply Makefile:124]: I was wrong, we have to keep as is, otherwise the name would be reflected in the generated files 👍

**Comment by sbueringer** [comment]
Thx!

/lgtm
/assign @fabriziopandini

**Comment by fabriziopandini** [comment]
@chrischdi great work!
After we got the first hooks let's audit together the resulting OpenAPI spec
/approve
/lgtm

**Pull request merged**